### PR TITLE
services/horizon: Implement horizon db fill-gaps command

### PR DESCRIPTION
**Pull request by tamirms**
<details>
  <summary>PR Checklist</summary>
  
### PR Structure

* [ ] This PR has reasonably narrow scope (if not, break it down into smaller PRs).
* [ ] This PR avoids mixing refactoring changes with feature changes (split into two PRs
  otherwise).
* [ ] This PR's title starts with name of package that is most changed in the PR, ex.
  `services/friendbot`, or `all` or `doc` if the changes are broad or impact many
  packages.

### Thoroughness

* [ ] This PR adds tests for the most critical parts of the new functionality or fixes.
* [ ] I've updated any docs ([developer docs](https://developers.stellar.org/api/), `.md`
  files, etc... affected by this change). Take a look in the `docs` folder for a given service,
  like [this one](https://github.com/stellar/go/tree/master/services/horizon/internal/docs).

### Release planning

* [ ] I've updated the relevant CHANGELOG ([here](services/horizon/CHANGELOG.md) for Horizon) if
  needed with deprecations, added features, breaking changes, and DB schema changes.
* [ ] I've decided if this PR requires a new major/minor version according to
  [semver](https://semver.org/), or if it's mainly a patch change. The PR is targeted at the next
  release branch if it's not a patch change.
</details>

### What

Close https://github.com/stellar/go/issues/4010

`horizon db fill-gaps` can be called with no parameters in which case it queries for any existing gaps in the horizon db and then proceeds to ingest history to fill the gaps.

The command can also be called with a start and end ledger parameter. In which case, the command will only query for gaps within the provided range.

`horizon db fill-gaps 1 1000` will fill any gaps occurring within the range 1-1000

It is also possible to configure the same parallel worker command line parameters that are used by horizon db reingest range so that fill-gaps can be also take advantage of concurrency.

### Known limitations

[N/A]

### CODE REVIEW

File: services/horizon/internal/db2/history/ledger.go [hunk 1]
```diff
@@ -130,20 +130,91 @@ func (q *Q) InsertLedger(ctx context.Context,
...
+			EndSequence:   min(end, oldestLedger-1),
+		})
+	}
+
+	gaps, err := q.GetLedgerGaps(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, gap := range gaps {
```
Comment (bartekn) [inline services/horizon/internal/db2/history/ledger.go:197]: I think this code depends on `gaps` being sorted. Are they always sorted?

File: services/horizon/internal/ingest/main.go [hunk 2]
```diff
@@ -486,23 +486,52 @@ func (s *system) VerifyRange(fromLedger, toLedger uint32, verifyState bool) erro
...
+		}
+		if from == 0 {
+			return errors.Errorf("Invalid range: %v genesis ledger starts at 1", pair)
+		}
+		if i == 0 {
+			continue
+		}
+		prev := ledgerRanges[i-1]
+		if prev[1] >= from {
+			return errors.Errorf("ranges are not sorted prevRange %v  curRange %v", prev, pair)
```
Comment (bartekn) [inline services/horizon/internal/ingest/main.go:503]:
```suggestion
			return errors.Errorf("ranges are not sorted prevRange %v curRange %v", prev, pair)
```

File: services/horizon/internal/ingest/parallel.go [hunk 3]
```diff
@@ -85,9 +104,17 @@ func calculateParallelLedgerBatchSize(rangeSize uint32, batchSizeSuggestion uint
 	return (batchSize / historyCheckpointLedgerInterval) * historyCheckpointLedgerInterval
 }
 
-func (ps *ParallelSystems) ReingestRange(fromLedger, toLedger uint32, batchSizeSuggestion uint32) error {
+func totalRangeSize(ledgerRanges [][2]uint32) uint32 {
+	var sum uint32
+	for _, pair := range ledgerRanges {
+		sum += pair[1] - pair[0]
```
Comment (bartekn) [inline services/horizon/internal/ingest/parallel.go:110]: Shouldn't we add +1 here? Like [1, 3] size is 3 (1, 2, 3) not 2.
```suggestion
		sum += pair[1] - pair[0] + 1
```

File: services/horizon/internal/db2/history/ledger.go (hunk 1 above)
Comment (tamirms) [reply services/horizon/internal/db2/history/ledger.go:197]: I think the sql query should return them in sorted order but I can add an explicit sort

**Comment by tamirms** [comment]
@bartekn can you take another look?

**Review by bartekn — APPROVED** [review]

**Pull request merged**